File: Array Diff/Array Diff.js


// Time Complexity: O(n + m), where n = arr1.length, m = arr2.length
function arrayDiff(arr1, arr2) {
    const set1 = new Set(arr1);
    const set2 = new Set(arr2);
    const diff = new Set([...set1].filter(x => !set2.has(x)).concat([...set2].filter(x => !set1.has(x))));
    return Array.from(diff).sort((a, b) => a - b);
}

// Input validation example
console.log(arrayDiff([1, 2, 3], [3, 4, 5]));

File: Array Diff/Array Diff.py


def array_diff(arr1, arr2):
    # Time Complexity: O(n + m), where n = len(arr1), m = len(arr2)
    # Space Complexity: O(n + m), for storing sets and result
    set1 = set(arr1)  # Convert arr1 to a set
    set2 = set(arr2)  # Convert arr2 to a set
    diff = set1.symmetric_difference(set2)  # Find symmetric difference
    return sorted(diff)  # Return sorted result

# Input validation example
if __name__ == "__main__":
    arr1 = [1, 2, 3]
    arr2 = [3, 4, 5]
    print(array_diff(arr1, arr2))
